lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+    if (typeof active_hooks_array[i][init_symbol] === 'function') {
+      runInitCallback(
+        active_hooks_array[i][init_symbol], asyncId, type, triggerId, resource);
+    }
+  }
+
+  // Isn't null if hooks were added/removed while the hooks were running.
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+function emitBeforeN(asyncId, triggerId) {
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][before_symbol] === 'function') {
+      runCallback(active_hooks_array[i][before_symbol], asyncId);
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// Usage: emitBeforeS(asyncId[, triggerId]). If triggerId is omitted then
+// asyncId will be used instead.
+function emitBeforeS(asyncId, triggerId) {
+  // CHECK(Number.isSafeInteger(asyncId) && asyncId > 0)
+  // CHECK(Number.isSafeInteger(triggerId) && triggerId > 0)
+
+  // Validate the ids.
+  if (asyncId < 0 || triggerId < 0) {
+    fatalError('before(): asyncId or triggerId is less than zero ' +
+               `(asyncId: ${asyncId}, triggerId: ${triggerId})`);
+  }
+
+  pushAsyncIds(asyncId, triggerId);
+
+  if (async_hook_fields[kBefore] === 0) {
+    return;
+  }
+
+  emitBeforeN(asyncId, triggerId);
+}
+
+
+// Called from native. The asyncId stack handling is taken care of there before
+// this is called.
+function emitAfterN(asyncId) {
+  if (async_hook_fields[kAfter] > 0) {
+    for (var i = 0; i < active_hooks_array.length; i++) {
+      if (typeof active_hooks_array[i][after_symbol] === 'function') {
+        runCallback(active_hooks_array[i][after_symbol], asyncId);
+      }
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// TODO(trevnorris): Calling emitBefore/emitAfter from native can't adjust the
+// kIdStackIndex. But what happens if the user doesn't have both before and
+// after callbacks.
+function emitAfterS(asyncId) {
+  emitAfterN(asyncId);
+  popAsyncIds(asyncId);
+}
+
+
+function emitDestroyS(asyncId) {
+  // Return early if there are no destroy callbacks, or on attempt to emit
+  // destroy on the void.
+  if (async_hook_fields[kDestroy] === 0 || asyncId === 0)
+    return;
+  async_wrap.addIdToDestroyList(asyncId);
+}
+
+
+function emitDestroyN(asyncId) {
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][destroy_symbol] === 'function') {
+      runCallback(active_hooks_array[i][destroy_symbol], asyncId);
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// Emit callbacks for native calls. Since some state can be setup directly from
+// C++ there's no need to perform all the work here.
+
+// This should only be called if hooks_array has kInit > 0. There are no global
+// values to setup. Though hooks_array will be cloned if C++ needed to visit
+// here since it's faster to do in JS.
```

Comment:
heh, horribly worded. It's saying if C++ needs to call the `init()` function below. Will reword.